### PR TITLE
Feedback api

### PR DESCRIPTION
**Pull request by patpragman**
so, I think I wired up the feedback API to a form, and a view.  Also, fixed a few bugs that were being weird in the API adjustment.  Now the duck is clickable and bounces around happily when you have the dialog open.

### CODE REVIEW

**Comment by schromya** [comment]
Seems to be a bug when a task is open:
Screenshot 2024-05-12 at 9 54 02 AM 1

**Comment by patpragman** [comment]
I'll look into it!

**Comment by patpragman** [comment]
Ok, @schromya - take a gander at it now, I think it's fixed, also, I added a way to see the projectID and the taskID in the panel, we can make that pretty later, but they should be tied to unique projects and tasks now.  Honestly, how I did this was pretty hacky and I'm not entirely sure it's "good idiomatic django" but I've got a book on the way for django development, so I'll read that then go through the entire app trying to find things that I can clean up in the next few days when it gets here.

**Comment by schromya** [comment]
Here's some UI things I noticed:
+ Another UI bug when screen size is reduced. Perhaps repositioning the modal and disallowing modal/duck on mobile would help?
Screenshot 2024-05-14 at 8 26 37 PM

+ The duck's behavior when the window is open is becoming overly distracting imo. Perhaps slowing it down would help?
+ Yeah having the task/project name instead of the ID would help.
+ I think it would be cool if the modal was a speech bubble positioned closer to the duck.

If you want, I can mess with some of the UI things :)

**Comment by schromya** [comment]
Also:

+ When the form is submitted, I think it would be intuitive if the formed closed.
+ Currently, no data is being saved to the db. If you sign up for Mongo Atlas, we can add you to our project so you can see the db.
+ Did we want to have the modal pop up randomly and not every time?

File: app/views.py [hunk 1]
```diff
@@ -64,9 +75,39 @@ def projectGraphView(request, projectID):
...
+def createTaskFeedback(request, projectID):
+
+    data, httpsCode = ProjectsAPIView.getProjects(
+        {"id": projectID}, getAuthorizationToken(request)
+    )
+
+    if request.method == "POST":
+        form = TaskFeedbackForm(request.POST)
+
+        if form.is_valid():
```
Comment (schromya) [inline app/views.py:102]: When the feedback is submitted, it's not making it into this `form.is_valid()` if-block.

**Comment by patpragman** [comment]
> Here's some UI things I noticed:
> 
> * Another UI bug when screen size is reduced. Perhaps repositioning the modal and disallowing modal/duck on mobile would help?
> 
> Screenshot 2024-05-14 at 8 26 37 PM
> * The duck's behavior when the window is open is becoming overly distracting imo. Perhaps slowing it down would help?
> * Yeah having the task/project name instead of the ID would help.
> * I think it would be cool if the modal was a speech bubble positioned closer to the duck.
> 
> If you want, I can mess with some of the UI things :)

Go for it!  You’re a better UI person than me.  Not sure how to fix the sizing thing, I’ll look into it.

File: app/views.py (hunk 1 above)
Comment (patpragman) [reply app/views.py:102]: I wonder why that is… I got my Django book yesterday and I’m reading through it, I’m not super sure how form validation works, I’ll look into it.

Comment (patpragman) [reply app/views.py:102]: Ok, I have context for this, I think I'm going to have to figure out how to pass the feedback form to the other views a little bit better.  Will advise.

**Comment by schromya** [comment]
Duck feedback complete. Appears 10% of the time.

File: api/models.py [hunk 2]
```diff
@@ -69,3 +70,14 @@ class Task(mongo.Document):
         "collection": "Task",  # Need to specify UPPER Case
         "strict": False,  # If true, throws weird error for __v
     }
+
+class Feedback(mongo.Document):
+    userID = mongo.ObjectIdField()
+    projectID = mongo.ObjectIdField()
```
Comment (crazykyler005) [inline api/models.py]: I would make the userID field required and possibly also the projectID as well

File: api/views/v1/feedback.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+                return { "message": "Need to have ONLY a id, taskID, or projectID property." }, status.HTTP_400_BAD_REQUEST
+            
+            feedback_objs = []
+
+            if "id" in feedbackData:
+                feedback_objs = Feedback.objects.get(id=feedbackData["id"])
+            elif "taskID" in feedbackData:
+                feedback_objs = Feedback.objects.filter(taskID__all=feedbackData["taskID"], **feedbackData)
+            elif "projectID" in feedbackData:
+                feedback_objs = Feedback.objects.filter(projectID__all=feedbackData["projectID"], **feedbackData)
```
Comment (crazykyler005) [inline api/views/v1/feedback.py]:
```suggestion
            if "id" in feedbackData:
                feedback_objs = Feedback.objects.get(id=feedbackData["id"])
            elif "taskID" in feedbackData:
                feedback_objs = Feedback.objects.filter(taskID=feedbackData["taskID"])
            elif "projectID" in feedbackData:
                feedback_objs = Feedback.objects.filter(projectID=feedbackData["projectID"])
```
I did some testing and it looks like the filter `taskID__all=feedbackData["taskID"]` doesn't actually work unless `feedbackData["taskID"]` is a list

File: api/views/v1/feedback.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        numberObjectsDeleted:int = 0
+
+        if "id" in feedbackData:
+            numberObjectsDeleted = Feedback.objects.get(id=feedbackData["id"]).delete()
+        elif "taskID" in feedbackData:
+            numberObjectsDeleted = Feedback.objects.filter(taskID__all=feedbackData["taskID"], **feedbackData).delete()
+        elif "projectID" in feedbackData:
+            numberObjectsDeleted = Feedback.objects.filter(projectID__all=feedbackData["projectID"], **feedbackData).delete()
+        elif "userID" in feedbackData:
+            numberObjectsDeleted = Feedback.objects.filter(userID__all=feedbackData["userID"], **feedbackData).delete()
```
Comment (crazykyler005) [inline api/views/v1/feedback.py]: same here:

```suggestion
        if "id" in feedbackData:
            numberObjectsDeleted = Feedback.objects.get(id=feedbackData["id"]).delete()
        elif "taskID" in feedbackData:
            numberObjectsDeleted = Feedback.objects.filter(taskID=feedbackData["taskID"]).delete()
        elif "projectID" in feedbackData:
            numberObjectsDeleted = Feedback.objects.filter(projectID__all=feedbackData["projectID"], **feedbackData).delete()
        elif "userID" in feedbackData:
            numberObjectsDeleted = Feedback.objects.filter(userID=feedbackData["userID"]).delete()
```

**Review by crazykyler005 — APPROVED** [review]
Looks good, just needs some api fixes done

**Comment by erwilliams64** [comment]
This is super cute, Melissa loves the duck! One concern is this message (shown in image) that pops up when you try to refresh after submitting. The default response should be Neutral and not "Very Unhappy".
image

**Comment by erwilliams64** [comment]
The duck also appeared when I had the edit project modal open and brought me to a white page which said "An error occurred: form not valid" image. I feel like the duck should work on any page

File: api/models.py (hunk 2 above)
Comment (schromya) [reply api/models.py]: Done.

**Comment by schromya** [comment]
> This is super cute, Melissa loves the duck! One concern is this message (shown in image) that pops up when you try to refresh after submitting. The default response should be Neutral and not "Very Unhappy". image

Done.

**Comment by schromya** [comment]
> The duck also appeared when I had the edit project modal open and brought me to a white page which said "An error occurred: form not valid" image. I feel like the duck should work on any page

Fixed.

**Pull request merged**